beluga/include/beluga/algorithm/estimation.hpp

```diff
@@ -35,6 +39,61 @@
 
 namespace beluga {
 
+namespace detail {
+
+/**
+ * \brief Given a range with quaternions and a normalized range of weights, compute the weighted average quaternion.
+ *
+ * \tparam QuaternionRange A [sized range](https://en.cppreference.com/w/cpp/ranges/sized_range) type whose
+ *  value type is `Eigen::Quaternion<Scalar>`.
+ * \tparam WeightsRange A [sized range](https://en.cppreference.com/w/cpp/ranges/sized_range) type whose
+ *  value type is `Scalar`.
+ * \tparam Scalar The scalar type, matching those of the weights.
+ * \param quaternions Range of quaternions to compute the weighted average from.
+ * \param normalized_weights Range of normalized weights. With matching indices wrt \c quaternions . Non-normalized
+ * weights will yield incorrect results. \return The average quaternion.
+ */
+template <
+    class QuaternionRange,
+    class WeightsRange,
+    class Scalar = typename ranges::range_value_t<QuaternionRange>::Scalar>
+Eigen::Quaternion<Scalar> weighted_average_quaternion(
+    const QuaternionRange& quaternions,
+    const WeightsRange& normalized_weights) {
+  // Implementation is based on https://github.com/strasdat/Sophus/blob/main/sophus/average.hpp#L133 with the variant of
+  // non-uniform weights.
+  const size_t num_quaternions = std::size(quaternions);
+  assert(num_quaternions >= 1);
+  Eigen::Matrix<Scalar, 4, Eigen::Dynamic> all_quaternions(4, num_quaternions);
+
+  int i = 0;
+  for (const auto& [quaternion, normalized_weight] : ranges::views::zip(quaternions, normalized_weights)) {
+    all_quaternions.col(i) = normalized_weight * quaternion.coeffs();
+    ++i;
+  }
+
+  Eigen::Matrix<Scalar, 4, 4> qqt = all_quaternions * all_quaternions.transpose();
+  Eigen::EigenSolver<Eigen::Matrix<Scalar, 4, 4>> es(qqt);
+
+  std::complex<Scalar> max_eigenvalue = es.eigenvalues()[0];
+  Eigen::Matrix<std::complex<Scalar>, 4, 1> max_eigenvector = es.eigenvectors().col(0);
+
+  for (int i = 1; i < 4; i++) {
+    if (std::norm(es.eigenvalues()[i]) > std::norm(max_eigenvalue)) {
+      max_eigenvalue = es.eigenvalues()[i];
+      max_eigenvector = es.eigenvectors().col(i);
+    }
+  }
+  Eigen::Quaternion<Scalar> quat;
+  quat.coeffs() <<                //
+      max_eigenvector[0].real(),  //
+      max_eigenvector[1].real(),  //
+      max_eigenvector[2].real(),  //
+      max_eigenvector[3].real();
```

Comment:
@serraramiro1 nit: I *think* `Eigen::Quaternion<Scalar>{max_eigenvector.real()}` should get you the same result.

Reply:
Not quite, added a coment